Zos3/web gateway

OmarElawady opened the pull request.

muhamadazmy committed:
initial commit

muhamadazmy committed:
scafolding for gateway daemon

OmarElawady committed:
wip: implement gateway nameproxy

OmarElawady committed:
fail non-public nodes gateway workloads even when trafik is running

muhamadazmy committed:
refactor yggdrasil networking to support movig

muhamadazmy committed:
fix tests

muhamadazmy committed:
support moving of yggdrasil to public namespace

Finish all the wiring need for the dynamic move

muhamadazmy committed:
wip: update node twin when ygg ip changes

muhamadazmy committed:
network stream public config

muhamadazmy commented on pkg/gateway/gateway.go:
```diff
@@ -0,0 +1,165 @@
+	configPath := filepath.Join(root, "proxy")
+	// where should service-restart/node-reboot recovery be handled?
+	if _, err := os.Stat(configPath); os.IsNotExist(err) {
+		err := os.MkdirAll(configPath, 0644)
```
MkdirAll is safe to call even if directory exists. the checks if it exists is a nice touch, but not needed.

muhamadazmy commented on pkg/gateway/gateway.go:
```diff
@@ -0,0 +1,165 @@
+		err := os.MkdirAll(configPath, 0644)
+		if err != nil {
+			// ok to panic?
+			panic(errors.Wrap(err, "couldn't make gateway config dir"))
```
Nope, yo need to return the error. change the New(root string) to `New(root string) (pkg.Gateway, error)

muhamadazmy commented on pkg/gateway/gateway.go:
```diff
@@ -0,0 +1,165 @@
+func isTraefikStarted() bool {
+	z, err := zinit.New("")
+	if err != nil {
+		panic(errors.Wrap(err, "couldn't get zinit client"))
```
No panicing please, always return an error. it's always the responsibility of the most top layer to decide either to exit on error or not.

muhamadazmy commented on pkg/gateway/gateway.go:
```diff
@@ -0,0 +1,165 @@
+		started = false
+	}
+	log.Debug().Str("state", traefikStatus.State.String()).Msg("checking traefik state")
+	started = traefikStatus.State.String() == zinit.ServiceStateRunning
```
```suggestion
	started = traefikStatus.State == zinit.ServiceStateRunning
```

OmarElawady replied:
the compiler complains with this form. There's `.Is` method in state.

muhamadazmy commented on pkg/primitives/gateway.go:
```diff
@@ -0,0 +1,84 @@
+	// gateway := stubs.NewGatewayStub(p.zbus)
+	// gateway.SetNamedProxy(ctx context.Context, arg0 string, arg1 []string)
+	gateway := stubs.NewGatewayStub(p.zbus)
+	if err := gateway.SetNamedProxy(ctx, fqdn, backends); err != nil {
```
I think an extra paramter (unique global workload id is needed to be passed here) so it's (ctx, wlId, fqdn, backends). the point is your config "name" must be the unique id. this way on decomission you just need to delete this config.

muhamadazmy commented on pkg/primitives/gateway.go:
```diff
@@ -0,0 +1,84 @@
+	if err := json.Unmarshal(wl.Data, &proxy); err != nil {
+		return fmt.Errorf("failed to unmarshal gateway proxy from reservation: %w", err)
+	}
+	if err := validateNameContract(twinID, proxy.Name); err != nil {
```
If we use the wrokdload id as a "key" to the domain configuration, this validation will not be needed. you just call the DeleteNamedProxy(wlid)

this is why we always use workload id to name all resources related to a workload. it's unique and is validated to be owned by the user (only owner can do decommission)

muhamadazmy commented on pkg/gateway/gateway.go:
```diff
@@ -0,0 +1,165 @@
+}
+
+func (g *gatewayModule) SetNamedProxy(fqdn string, backends []string) error {
+	if !namespace.Exists(types.PublicNamespace) {
```
I already added a method to do this check (but available only on my network changes branch) for now. create a method 
```
func (g *gatewayModule) isGatewaySupported()
```
that does this call. which then can be modified later
```

muhamadazmy commented on pkg/gateway/gateway.go:
```diff
@@ -0,0 +1,165 @@
+	return nil
+}
+func (g *gatewayModule) DeleteNamedProxy(fqdn string) error {
+	filename := filepath.Join(g.proxyConfigPath, fmt.Sprintf("%s.yaml", fqdn))
```
You need to pass the workload id to both Set and Delete method, and use this workload id as a granteed unique file name. also it is easy to reverse the config file to an actual deployment.

muhamadazmy committed:
rewire yggdrasil is complete

- node monitor changes
- setting of public config will trigger the rewiring
- changes will reflect in both the node and the twin object

muhamadazmy committed:
Merge pull request #1373 from threefoldtech/zos3/web-gateway-network

web gateway network

muhamadazmy committed:
persist public config

muhamadazmy committed:
rename error

OmarElawady committed:
add result - no panic - use workload id

OmarElawady committed:
remove hardcoded domain

muhamadazmy committed:
Merge branch 'zos3/web-gateway' of github.com:threefoldtech/zos into zos3/web-gateway

muhamadazmy committed:
some minor refacoting

muhamadazmy committed:
static config for traefik

muhamadazmy committed:
Get traefik from an flist

muhamadazmy committed:
updating go version to 1.17

muhamadazmy committed:
ci fixes
